File: core/common/src/androidMain/kotlin/br/com/ricarlo/common/PerformanceProviderImpl.kt
package br.com.ricarlo.common

import com.google.firebase.perf.FirebasePerformance
import com.google.firebase.perf.trace

internal class PerformanceProviderImpl : PerformanceProvider {

    private val performance: FirebasePerformance by lazy {
        FirebasePerformance.getInstance()
    }

    override fun <T> trace(name: String, action: () -> T): T {
        if (!performance.isPerformanceCollectionEnabled) {
            return action()
        }

        return performance.newTrace(name).trace {
            action()
        }
    }

    override fun setPerformanceCollectionEnabled(enabled: Boolean) {
        performance.isPerformanceCollectionEnabled = enabled
    }
}

File: iosApp/iosApp/PerformanceProviderImpl.swift
//
//  Created by Ricarlo Silva on 22/09/25.
//  Copyright © 2025 orgName. All rights reserved.
//
import FirebasePerformance
import shared

internal class PerformanceProviderImpl: PerformanceProvider {

    private lazy var performance: Performance = {
        Performance.sharedInstance()
    }()

    func trace(name: String, action: @escaping () -> Any?) -> Any? {
        if !performance.isDataCollectionEnabled {
            return action()
        }
        guard let trace = Performance.startTrace(name: name) else {
            return action()
        }
        let result = action()
        trace.stop()
        return result
    }

    func setPerformanceCollectionEnabled(enabled: Bool) {
        performance.isDataCollectionEnabled = enabled
    }
}
